addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -1267,11 +1283,301 @@ export class OdooEditor extends EventTarget {
+                // using the fact that the last element of contentsElements corresponds to docEnd or is one of its children
+                let docE = docEnd;
+                let fragE = contentsElements[contentsElements.length - 1];
+                while (fragE && fragE.oid && fragE.parentElement) {
+                    // fragE could have been fully selected (and thus is a child of docE)
+                    // fragE cannot have an oid if it is still present in the document.
+                    fragE = fragE.parentElement;
+                }
+                while (fragE && fragE.parentElement) {
+                    // docE and fragE are synchronized, reducing depth until fragE has no parent (its parent would then be this.editable)
+                    docE = docE.parentElement;
+                    fragE = fragE.parentElement;
+                }
+                docEndRoot = docE;
+                docStartRoot = docE.previousElementSibling || docEndRoot;
+            }
+
+            // extract children of root from an array of unremovables which is ordered and in reverse order.
+            // children of root are guaranteed to be at the end of unremovables if any.
+            // to maintain O(n) complexity, push and reverse are used instead of shift
+            const extractUnremovableChildren = function (root, unremovables) {
+                const unremovableChildren = [];
+                while (unremovables.length) {
+                    const unremovable = unremovables.pop();
+                    if (root.contains(unremovable)) {
+                        unremovableChildren.push(unremovable);
+                    } else {
+                        unremovables.push(unremovable);
+                        break;
+                    }
+                }
+                return unremovableChildren.reverse();
+            };
+
+            // used to fill nodes with an empty <p><br></p> (i.e. when emptying an contentEditable=True)
+            const ensureEmptyPElement = function (nodes) {
+                if (!nodes.length) {
+                    const pElement = this.document.createElement('P');
+                    const brElement = this.document.createElement('BR');
+                    pElement.append(brElement);
+                    nodes.push(pElement);
+                }
+            }.bind(this);
+
+            // main algorithm used to restore unremovables which were completely removed from document.
+            // unremovables is an array of unremovable stored in reverse traversal order to maintain O(n) complexity
+            // since it has to be completely emptied to finish the work.
+            // destination is an array which will store the result of the restoration, and contains nodes to be added
+            // to document in order.
+            // editableContext is the value that isContentEditable should have when contentEditable is 'inherit', when
+            // the fragment is not able to provide inheritance.
```

Comment:
I appreciate that you wrote a lot of comments in this PR, thanks ! 👍  That being said, could you please update them so that they start with a capital letter, end with proper punctuation (you got this part right already) and do not exceed a threshold of 80 characters ? That would be perfect, thanks ! :)